src/rgw/rgw_datalog.cc

```diff
@@ -898,19 +910,22 @@ void RGWDataChangesLog::renew_stop()
   renew_cond.notify_all();
 }
 
-void RGWDataChangesLog::mark_modified(int shard_id, const rgw_bucket_shard& bs)
+void RGWDataChangesLog::mark_modified(int shard_id, const rgw_bucket_shard& bs, uint64_t gen)
 {
-  auto key = bs.get_key();
+  const auto key = bs.get_key();
   {
     std::shared_lock rl{modified_lock}; // read lock to check for existence
-    auto shard = modified_shards.find(shard_id);
-    if (shard != modified_shards.end() && shard->second.count(key)) {
-      return;
+    const auto& shard = modified_shards.find(shard_id);
+    if (shard != modified_shards.end()) {
+      for (const auto& [k, g] : shard->second) {
+        if (k == key && g == gen)
```

Comment:
set is an ordered container, so we should prefer binary searches to linear ones. does `shard->second.count({key, gen})` work here?

Reply:
yes, it should work.